### ISSUE TITLE
[BUG] Restoring databases with dot in the name sometimes fails with `Error switching to database mydumper_0 Restoring table structure`

### ISSUE THREAD

**Issue by NavidSassan** [mydumper/mydumper#1307]
Restoring databases with dot in the name sometimes fails with `Error switching to database mydumper_0 Restoring table structure`.

**To Reproduce**
```bash
dnf install mariadb-server wget zstd -y
dnf install https://github.com/mydumper/mydumper/releases/download/v0.15.1-3/mydumper-0.15.1-3.el8.x86_64.rpm -y
systemctl start mariadb

wget https://downloads.mysql.com/docs/world_x-db.tar.gz
tar xvzf world_x-db.tar.gz
sed --in-place 's/world_x/world.x/g' world_x-db/world_x.sql

mysql < world_x-db/world_x.sql # errors do not matter here, we just want some tables in the db

for i in {1..15}; do
    echo "Run $i"
    \rm -rf mariadb-dump/; mkdir mariadb-dump
    mydumper --outputdir mariadb-dump --logfile mariadb-dump/mydumper.log --verbose 3 --compress ZSTD
    myloader --directory=mariadb-dump --overwrite-tables
    echo -e "Return Code $?\n"
    sleep 1
done
```
* Around 3 runs fail with `Error switching to database mydumper_0 Restoring table structure`
* Only seems to happen with `--compress ZSTD`
* The problem seems to happen during the backup (not the restore). The restore of a given backup either consistenly works, or consistenly fails. As in, running `myloader` repeatedly on a "broken" backup always fails.

**Expected behavior**
The restore of the backup should always work.

**Log**
[mydumper-errors.log](https://github.com/mydumper/mydumper/files/12701811/mydumper-errors.log)
[myloader-errors.log](https://github.com/mydumper/mydumper/files/12701813/myloader-errors.log)

[mydumper-working.log](https://github.com/mydumper/mydumper/files/12701812/mydumper-working.log)
[myloader-working.log](https://github.com/mydumper/mydumper/files/12701814/myloader-working.log)

**Backup**
If applicable, add the list of content of the database path, an `ls -l` will be enough
```bash
ll mariadb-dump-zstd-errors/
total 228
-rw-r--r--. 1 root root  1624 Sep 22 14:06 metadata
-rw-r--r--. 1 root root 77114 Sep 22 14:06 mydumper_0.city.00000.sql.zst
-rw-r--r--. 1 root root   313 Sep 22 14:06 mydumper_0.city-schema.sql.zst
-rw-r--r--. 1 root root  3983 Sep 22 14:06 mydumper_0.country.00000.sql.zst
-rw-r--r--. 1 root root   255 Sep 22 14:06 mydumper_0.country-schema.sql.zst
-rw-r--r--. 1 root root   126 Sep 22 14:06 mydumper_0-schema-create.sql.zst
-rw-r--r--. 1 root root    13 Sep 22 14:06 mydumper_0-schema-triggers.sql.zst
-rw-r--r--. 1 root root 15979 Sep 22 14:06 mydumper.log
-rw-r--r--. 1 root root   465 Sep 22 14:06 mysql.db-schema.sql.zst
-rw-r--r--. 1 root root  1080 Sep 22 14:06 mysql.event-schema.sql.zst
-rw-r--r--. 1 root root   319 Sep 22 14:06 mysql.func-schema.sql.zst
-rw-r--r--. 1 root root   307 Sep 22 14:06 mysql.help_category-schema.sql.zst
-rw-r--r--. 1 root root   279 Sep 22 14:06 mysql.help_keyword-schema.sql.zst
-rw-r--r--. 1 root root   259 Sep 22 14:06 mysql.help_relation-schema.sql.zst
-rw-r--r--. 1 root root   328 Sep 22 14:06 mysql.help_topic-schema.sql.zst
-rw-r--r--. 1 root root   318 Sep 22 14:06 mysql.index_stats-schema.sql.zst
-rw-r--r--. 1 root root   244 Sep 22 14:06 mysql.plugin-schema.sql.zst
-rw-r--r--. 1 root root   569 Sep 22 14:06 mysql.proc.00000.sql.zst
-rw-r--r--. 1 root root  1008 Sep 22 14:06 mysql.proc-schema.sql.zst
-rw-r--r--. 1 root root   450 Sep 22 14:06 mysql.procs_priv-schema.sql.zst
-rw-r--r--. 1 root root   199 Sep 22 14:06 mysql.proxies_priv.00000.sql.zst
-rw-r--r--. 1 root root   363 Sep 22 14:06 mysql.proxies_priv-schema.sql.zst
-rw-r--r--. 1 root root   310 Sep 22 14:06 mysql.roles_mapping-schema.sql.zst
-rw-r--r--. 1 root root   121 Sep 22 14:06 mysql-schema-create.sql.zst
-rw-r--r--. 1 root root    13 Sep 22 14:06 mysql-schema-triggers.sql.zst
-rw-r--r--. 1 root root   337 Sep 22 14:06 mysql.servers-schema.sql.zst
-rw-r--r--. 1 root root   476 Sep 22 14:06 mysql.tables_priv-schema.sql.zst
-rw-r--r--. 1 root root   280 Sep 22 14:06 mysql.table_stats-schema.sql.zst
-rw-r--r--. 1 root root   289 Sep 22 14:06 mysql.time_zone_leap_second-schema.sql.zst
-rw-r--r--. 1 root root   263 Sep 22 14:06 mysql.time_zone_name-schema.sql.zst
-rw-r--r--. 1 root root   286 Sep 22 14:06 mysql.time_zone-schema.sql.zst
-rw-r--r--. 1 root root   283 Sep 22 14:06 mysql.time_zone_transition-schema.sql.zst
-rw-r--r--. 1 root root   327 Sep 22 14:06 mysql.time_zone_transition_type-schema.sql.zst
-rw-r--r--. 1 root root   402 Sep 22 14:06 mysql.transaction_registry-schema.sql.zst
-rw-r--r--. 1 root root   208 Sep 22 14:06 mysql.user.00000.sql.zst
-rw-r--r--. 1 root root   794 Sep 22 14:06 mysql.user-schema.sql.zst
```
```bash
ll mariadb-dump-zstd-working/
total 228
-rw-r--r--. 1 root root  1624 Sep 22 14:05 metadata
-rw-r--r--. 1 root root 77114 Sep 22 14:05 mydumper_0.city.00000.sql.zst
-rw-r--r--. 1 root root   313 Sep 22 14:05 mydumper_0.city-schema.sql.zst
-rw-r--r--. 1 root root  3983 Sep 22 14:05 mydumper_0.country.00000.sql.zst
-rw-r--r--. 1 root root   255 Sep 22 14:05 mydumper_0.country-schema.sql.zst
-rw-r--r--. 1 root root   126 Sep 22 14:05 mydumper_0-schema-create.sql.zst
-rw-r--r--. 1 root root    13 Sep 22 14:05 mydumper_0-schema-triggers.sql.zst
-rw-r--r--. 1 root root 15979 Sep 22 14:05 mydumper.log
-rw-r--r--. 1 root root   465 Sep 22 14:05 mysql.db-schema.sql.zst
-rw-r--r--. 1 root root  1080 Sep 22 14:05 mysql.event-schema.sql.zst
-rw-r--r--. 1 root root   319 Sep 22 14:05 mysql.func-schema.sql.zst
-rw-r--r--. 1 root root   307 Sep 22 14:05 mysql.help_category-schema.sql.zst
-rw-r--r--. 1 root root   279 Sep 22 14:05 mysql.help_keyword-schema.sql.zst
-rw-r--r--. 1 root root   259 Sep 22 14:05 mysql.help_relation-schema.sql.zst
-rw-r--r--. 1 root root   328 Sep 22 14:05 mysql.help_topic-schema.sql.zst
-rw-r--r--. 1 root root   318 Sep 22 14:05 mysql.index_stats-schema.sql.zst
-rw-r--r--. 1 root root   244 Sep 22 14:05 mysql.plugin-schema.sql.zst
-rw-r--r--. 1 root root   569 Sep 22 14:05 mysql.proc.00000.sql.zst
-rw-r--r--. 1 root root  1008 Sep 22 14:05 mysql.proc-schema.sql.zst
-rw-r--r--. 1 root root   450 Sep 22 14:05 mysql.procs_priv-schema.sql.zst
-rw-r--r--. 1 root root   199 Sep 22 14:05 mysql.proxies_priv.00000.sql.zst
-rw-r--r--. 1 root root   363 Sep 22 14:05 mysql.proxies_priv-schema.sql.zst
-rw-r--r--. 1 root root   310 Sep 22 14:05 mysql.roles_mapping-schema.sql.zst
-rw-r--r--. 1 root root   121 Sep 22 14:05 mysql-schema-create.sql.zst
-rw-r--r--. 1 root root    13 Sep 22 14:05 mysql-schema-triggers.sql.zst
-rw-r--r--. 1 root root   337 Sep 22 14:05 mysql.servers-schema.sql.zst
-rw-r--r--. 1 root root   476 Sep 22 14:05 mysql.tables_priv-schema.sql.zst
-rw-r--r--. 1 root root   280 Sep 22 14:05 mysql.table_stats-schema.sql.zst
-rw-r--r--. 1 root root   289 Sep 22 14:05 mysql.time_zone_leap_second-schema.sql.zst
-rw-r--r--. 1 root root   263 Sep 22 14:05 mysql.time_zone_name-schema.sql.zst
-rw-r--r--. 1 root root   286 Sep 22 14:05 mysql.time_zone-schema.sql.zst
-rw-r--r--. 1 root root   283 Sep 22 14:05 mysql.time_zone_transition-schema.sql.zst
-rw-r--r--. 1 root root   327 Sep 22 14:05 mysql.time_zone_transition_type-schema.sql.zst
-rw-r--r--. 1 root root   402 Sep 22 14:05 mysql.transaction_registry-schema.sql.zst
-rw-r--r--. 1 root root   208 Sep 22 14:05 mysql.user.00000.sql.zst
-rw-r--r--. 1 root root   794 Sep 22 14:05 mysql.user-schema.sql.zst
```

```bash
diff -rq mariadb-dump-zstd-errors/ mariadb-dump-zstd-working/ | sort
# Files mariadb-dump-zstd-errors/metadata and mariadb-dump-zstd-working/metadata differ
# Files mariadb-dump-zstd-errors/mydumper.log and mariadb-dump-zstd-working/mydumper.log differ
# Only in mariadb-dump-zstd-errors/: mydumper_0.country-schema.sql


diff mariadb-dump-zstd-errors/metadata mariadb-dump-zstd-working/metadata
# 1c1
# < # Started dump at: 2023-09-22 14:06:05
# ---
# > # Started dump at: 2023-09-22 14:05:53
# 102c102
# < # Finished dump at: 2023-09-22 14:06:05
# ---
# > # Finished dump at: 2023-09-22 14:05:53
```

**Environment**
 - OS version: Rocky Linux 8.8
```bash
mydumper --version
# mydumper v0.15.1-3, built against MySQL 5.7.42-46 with SSL support

myloader --version
# myloader v0.15.1-3, built against MySQL 5.7.42-46 with SSL support

zstd --version
# *** zstd command line interface 64-bits v1.4.4, by Yann Collet ***

/usr/libexec/mysqld --version
# /usr/libexec/mysqld  Ver 10.3.39-MariaDB for Linux on x86_64 (MariaDB Server)
```

Let me know if there is anything else I can test / help with.

**Comment by NavidSassan** [comment]
After some `gdb` debugging, I think I know what's causing the problem, but I'm not sure how to actually fix it (note: all debugging was done on 8b78ef2):

* at first there is a "wrong" entry in `db_hash`, mapping `mydumper_0` (filename) to `mydumper_0` (real_database)
* as soon as `process_database_filename()` runs, it fixes the entry in `db_hash` (`mydumper_0` -> `world.x`, in my scenario above)
* however, `process_table_filename()` runs before `process_database_filename()`, therefore it gets the wrong entry, leading to the error message


I tried to verify this by manually setting the value during a `gdb` run, which worked:
```bash
gdb --args ./myloader --user=root --host 127.0.0.1 --password=password --directory=/tmp/mariadb-dump --overwrite-tables --verbose 4 --threads 1
(gdb) b myloader_process.c:445 if $_regex(filename, "mydumper_.*")

Thread 12 "myloader" hit Breakpoint 1, process_table_filename (filename=0x47f600 "mydumper_0.city-schema.sql")
    at /home/navid.sassan/git/other/mydumper/src/myloader_process.c:447
447       dbt=append_new_db_table(real_db_name, table_name,0,NULL);

(gdb) p real_db_name->real_database
$1 = 0x7fffc800a790 "mydumper_0"

(gdb) set real_db_name->real_database="world.x"

(gdb) p real_db_name->real_database
$2 = 0x7fffc80149e0 "world.x"

(gdb) c
Continuing.
```
This run fails with `Error switching to database `mydumper_0` Restoring post tasks`, but it gets further then the error described above. It manages to restore the tables (including their content) of the database, which failed before.

I also tried to isolate the commit that introduced this bug using `git bisect run` with the script below. This resulted in
```
# possible first bad commit: [4e5b7ff55a332855ea960a8b0090759cc66cc103] [ Unifying ] Directory using intermediate queue as Stream
# possible first bad commit: [e10dd684013ea2bad0c24cc5be1ee8da6ea42a09] [ Unifying ] Directory using intermediate queue as Stream
```
* 4e5b7ff55a332855ea960a8b0090759cc66cc103 always results in a segfault
* e10dd684013ea2bad0c24cc5be1ee8da6ea42a09 fails to compile

but:
* b02bffe (commit after 4e5b7ff) results in the same error as v0.15.1-3 (`Error switching to database mydumper_0 Restoring tables (1)`)
* 025de0f (commit before e10dd68) always manages to restore the database with all the tables

["git bisect run" script](https://github.com/mydumper/mydumper/files/12832687/test.sh.md)


I really hope this helps to implement a fix for the problem. If there is any other scenario I should try please let me know. Thanks a lot

**Comment by davidducos** [comment]
Hi @NavidSassan,

Your issue was not as simple as I thought, but I was able to find the the mistake.
